Задание 1

In [1]:
import pandas as pd

In [2]:
horses = pd.DataFrame()

with open('horse_data.csv', 'r', encoding='utf-8') as f:
    for line in f:
        if line.split(',')[1] == '1':
            Age = 'Adult horse'
        elif line.split(',')[1] == '2':
            Age = 'Young (< 6 months)'
        else:
            Age = 'No information'
            
        if line.split(',')[12] == '1':
            distension = 'none'
        elif line.split(',')[12] == '2':
            distension = 'slight'
        elif line.split(',')[12] == '3':
            distension = 'moderate'
        elif line.split(',')[12] == '4':
            distension = 'severe'
        else:
            distension = 'No information'

        if line.split(',')[16] == '1':
            examination = 'normal'
        elif line.split(',')[16] == '2':
            examination = 'increased'
        elif line.split(',')[16] == '3':
            examination = 'decreased'
        elif line.split(',')[16] == '4':
            examination = 'absent'
        else:
            examination = 'No information'
            
        if line.split(',')[17] == '1':
            abdomen = 'normal'
        elif line.split(',')[17] == '2':
            abdomen = 'other'
        elif line.split(',')[17] == '3':
            abdomen = 'firm feces in the large intestine'
        elif line.split(',')[17] == '4':
            abdomen = 'distended small intestine'
        elif line.split(',')[17] == '5':
            abdomen = 'distended large intestine'
        else:
            abdomen = 'No information'
            
        if line.split(',')[20] == '1':
            appearance = 'clear'
        elif line.split(',')[20] == '2':
            appearance = 'cloudy'
        elif line.split(',')[20] == '3':
            appearance = 'serosanguinous'
        else:
            appearance = 'No information'
            
        if line.split(',')[3] == '?':
            temperature = None
        else:
            temperature = float(line.split(',')[3])
 
        if line.split(',')[21] == '?':
            protein = None
        else:
            protein = float(line.split(',')[21])
        
        row = {'Age': Age, 
               'Hospital Number': line.split(',')[2], 
               'rectal temperature': temperature, 
               'abdominal distension': distension, 
               'rectal examination': examination, 
               'abdomen': abdomen, 
               'abdominocentesis appearance': appearance, 
               'abdomcentesis total protein': protein}
        
        horses = pd.concat([horses, pd.DataFrame([row])])
        
# horses.head(15)
horses

Age Hospital Number  rectal temperature abdominal distension  \
0      Adult horse          530101                38.5               severe   
0      Adult horse          534817                39.2               slight   
0      Adult horse          530334                38.3                 none   
0   No information         5290409                39.1               severe   
0      Adult horse          530255                37.3       No information   
..             ...             ...                 ...                  ...   
0      Adult horse          533886                 NaN       No information   
0      Adult horse          527702                37.2             moderate   
0      Adult horse          529386                37.5             moderate   
0      Adult horse          530612                36.5             moderate   
0      Adult horse          534618                37.2       No information   

   rectal examination                    abdomen abdominocentesis appearance  \
0           decreased  distended large intestine              No information   
0              absent                      other                      cloudy   
0              normal                     normal              No information   
0           decreased             No information              serosanguinous   
0      No information             No information              No information   
..                ...                        ...                         ...   
0      No information  distended large intestine              No information   
0              absent  distended small intestine              serosanguinous   
0           decreased  distended large intestine              No information   
0              absent  distended small intestine              serosanguinous   
0              absent                     normal                       clear   

   abdomcentesis total protein  
0                         None  
0                            2  
0                         None  
0                          5.3  
0                         None  
..                         ...  
0                         None  
0                          3.3  
0                         None  
0                          3.4  
0                            1  

[300 rows x 8 columns]

In [3]:
horses['Age'].describe()

count             300
unique              2
top       Adult horse
freq              276
Name: Age, dtype: object

In [4]:
horses['rectal temperature'].describe()

count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: rectal temperature, dtype: float64

In [5]:
horses['abdominal distension'].describe()

count      300
unique       5
top       none
freq        76
Name: abdominal distension, dtype: object

In [6]:
horses['abdomen'].describe()

count                300
unique                 6
top       No information
freq                 118
Name: abdomen, dtype: object

In [7]:
horses['abdominocentesis appearance'].describe()

count                300
unique                 4
top       No information
freq                 165
Name: abdominocentesis appearance, dtype: object

In [8]:
horses['rectal examination'].describe()

count                300
unique                 5
top       No information
freq                 102
Name: rectal examination, dtype: object

In [9]:
horses['abdomcentesis total protein'].describe()

count     102.0
unique     37.0
top         2.0
freq       26.0
Name: abdomcentesis total protein, dtype: float64

Задание 2. Выбросы 


In [10]:
q1 = horses['rectal temperature'].quantile(0.25)
q3 = horses['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horses[horses['rectal temperature'].between(lower_bound, upper_bound, inclusive=True)].sort_values('rectal temperature')
remove_outliers

Age Hospital Number  rectal temperature abdominal distension  \
0      Adult horse          527933                36.8       No information   
0      Adult horse          528620                36.9               slight   
0      Adult horse          528214                37.0             moderate   
0      Adult horse          530034                37.0             moderate   
0      Adult horse          528742                37.1                 none   
..             ...             ...                 ...                  ...   
0      Adult horse          534925                39.4             moderate   
0      Adult horse          530028                39.5                 none   
0   No information         5305129                39.5       No information   
0      Adult horse         5290482                39.5             moderate   
0      Adult horse          535176                39.5               slight   

   rectal examination                            abdomen  \
0      No information                     No information   
0      No information                     No information   
0              absent          distended small intestine   
0              normal          distended large intestine   
0           decreased  firm feces in the large intestine   
..                ...                                ...   
0              absent                     No information   
0              absent                     No information   
0      No information                     No information   
0              absent          distended large intestine   
0           decreased                     No information   

   abdominocentesis appearance abdomcentesis total protein  
0               No information                          10  
0               No information                        None  
0               No information                        None  
0                       cloudy                        None  
0               serosanguinous                        None  
..                         ...                         ...  
0                        clear                           2  
0               No information                         3.6  
0               No information                        None  
0                        clear                        None  
0                        clear                        None  

[226 rows x 8 columns]

В выбросы попали лошади с температурой 36.7 и ниже и 39.55 и выше. При этом нормальная ректальная температура лошади - от 37,5 до 38,5. То есть в выбросы попали животные со значительным отклонением от нормы. Следовательно, их надо срочно лечить, если они еще живы. 

То есть исчисления помогли нам прикинуть, сколько у нас здоровых лошадей, а сколько мы пока не можем использовать.

In [11]:
q1 = horses['abdomcentesis total protein'].quantile(0.25)
q3 = horses['abdomcentesis total protein'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = horses[horses['abdomcentesis total protein'].between(lower_bound, upper_bound, inclusive=True)].sort_values('abdomcentesis total protein')
remove_outliers

Age Hospital Number  rectal temperature abdominal distension  \
0   No information         5288249                39.0       No information   
0      Adult horse          534618                37.2       No information   
0      Adult horse          534579                38.7                 none   
0      Adult horse         5262541                37.4                 none   
0      Adult horse          535330                37.4             moderate   
..             ...             ...                 ...                  ...   
0      Adult horse          529126                38.0               slight   
0   No information         5290409                39.1               severe   
0      Adult horse          527677                39.0             moderate   
0      Adult horse          528630                37.8       No information   
0      Adult horse          529663                38.2       No information   

   rectal examination                    abdomen abdominocentesis appearance  \
0      No information             No information              No information   
0              absent                     normal                       clear   
0              normal             No information                       clear   
0              normal             No information                      cloudy   
0      No information             No information              No information   
..                ...                        ...                         ...   
0           decreased             No information                       clear   
0           decreased             No information              serosanguinous   
0      No information             No information              serosanguinous   
0              absent             No information                       clear   
0              absent  distended small intestine                      cloudy   

   abdomcentesis total protein  
0                          0.1  
0                            1  
0                            1  
0                            1  
0                            1  
..                         ...  
0                          5.3  
0                          5.3  
0                            6  
0                            6  
0                          6.6  

[95 rows x 8 columns]

Что касается показателя "abdomcentesis total protein", по нему есть данные только о 102 особях. Кажется, это очень мало, чтобы делать какие-то выводы. 


Задание 3. Пропуски


In [12]:
len(horses[horses['Age'] == 'No information'])

24

In [13]:
horses[horses['Age'] == 'No information'].head(24)

Age Hospital Number  rectal temperature abdominal distension  \
0  No information         5290409                39.1               severe   
0  No information         5299629                38.3                 none   
0  No information         5291329                38.0               slight   
0  No information         5301219                 NaN             moderate   
0  No information          534998                38.3       No information   
0  No information         5277409                39.2       No information   
0  No information         5288249                39.0       No information   
0  No information         5282839                38.6               severe   
0  No information         5292929                 NaN       No information   
0  No information          534092                39.7       No information   
0  No information         5290759                38.1                 none   
0  No information         5292489                38.0               slight   
0  No information         5297379                38.4             moderate   
0  No information         5287179                38.1             moderate   
0  No information         5287279                 NaN                 none   
0  No information         5294369                38.9             moderate   
0  No information         5289419                 NaN       No information   
0  No information          534597                38.5       No information   
0  No information         5291409                38.2               severe   
0  No information         5291719                38.8               severe   
0  No information         5294539                38.8             moderate   
0  No information         5305629                38.6       No information   
0  No information         5297159                38.8               slight   
0  No information         5305129                39.5       No information   

  rectal examination                    abdomen abdominocentesis appearance  \
0          decreased             No information              serosanguinous   
0          decreased             No information                       clear   
0             normal                     normal                       clear   
0             absent  distended large intestine              serosanguinous   
0     No information             No information              No information   
0     No information             No information              No information   
0     No information             No information              No information   
0             absent             No information                      cloudy   
0     No information             No information              No information   
0     No information             No information                      cloudy   
0             normal             No information              No information   
0          increased                     normal              No information   
0     No information             No information                      cloudy   
0          increased             No information                      cloudy   
0             normal             No information              No information   
0     No information             No information                       clear   
0     No information             No information                      cloudy   
0             normal             No information                       clear   
0     No information             No information                       clear   
0             absent  distended small intestine              serosanguinous   
0          increased             No information              No information   
0     No information             No information              No information   
0     No information             No information              No information   
0     No information             No information              No information   

  abdomcentesis total protein  
0                         5.3  
0                      

Скорее всего, в дата-сете речь идет об "учебных экземплярах" - то есть заведомо взрослых лошадях. Поэтому пропуски в столбце Возраст можно заменить на "Adult horse".


In [14]:
horses['rectal temperature'].describe()

count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: rectal temperature, dtype: float64

В столбце rectal temperature - 60 пропусков. Заменить их средним значением будет некорректно - вдруг у лошади аномальная температура, и ее надо лечить. Следовательно, можно удалить строки с отсутствующей температурой из дата-сета - до появления показаний.

In [15]:
len(horses[horses['abdominal distension'] == 'No information'])

56

In [16]:
len(horses[horses['abdomen'] == 'No information'])

118

In [17]:
len(horses[horses['abdominocentesis appearance'] == 'No information'])

165

In [18]:
len(horses[horses['rectal examination'] == 'No information'])

102

In [19]:
len(horses) - len(horses[horses['abdomcentesis total protein'] > 0])

198

In [20]:
horses = pd.DataFrame()

with open('horse_data.csv', 'r', encoding='utf-8') as f:
    for line in f:
            
        if line.split(',')[12] == '1':
            distension = 'none'
        elif line.split(',')[12] == '2':
            distension = 'slight'
        elif line.split(',')[12] == '3':
            distension = 'moderate'
        elif line.split(',')[12] == '4':
            distension = 'severe'
        else:
            distension = 'No information'

        if line.split(',')[16] == '1':
            examination = 'normal'
        elif line.split(',')[16] == '2':
            examination = 'increased'
        elif line.split(',')[16] == '3':
            examination = 'decreased'
        elif line.split(',')[16] == '4':
            examination = 'absent'
        else:
            examination = 'No information'
            
        if line.split(',')[17] == '1':
            abdomen = 'normal'
        elif line.split(',')[17] == '2':
            abdomen = 'other'
        elif line.split(',')[17] == '3':
            abdomen = 'firm feces in the large intestine'
        elif line.split(',')[17] == '4':
            abdomen = 'distended small intestine'
        elif line.split(',')[17] == '5':
            abdomen = 'distended large intestine'
        else:
            abdomen = 'No information'
            
        if line.split(',')[20] == '1':
            appearance = 'clear'
        elif line.split(',')[20] == '2':
            appearance = 'cloudy'
        elif line.split(',')[20] == '3':
            appearance = 'serosanguinous'
        else:
            appearance = 'No information'
            
        if line.split(',')[3] == '?':
            temperature = None
        else:
            temperature = float(line.split(',')[3])
 
        if line.split(',')[21] == '?':
            protein = None
        else:
            protein = float(line.split(',')[21])
        
        row = {'Age': 'Adult horse', 
               'Hospital Number': line.split(',')[2], 
               'rectal temperature': temperature, 
               'abdominal distension': distension, 
               'rectal examination': examination, 
               'abdomen': abdomen, 
               'abdominocentesis appearance': appearance, 
               'abdomcentesis total protein': protein}
        
        horses = pd.concat([horses, pd.DataFrame([row])])

In [21]:
horses[(horses['rectal temperature'] > 0) 
   & (horses['abdominal distension'] != 'No information') 
   & (horses['abdomen'] != 'No information') 
   & (horses['abdominocentesis appearance'] != 'No information') 
   & (horses['rectal examination'] != 'No information')
   & (horses['abdomcentesis total protein'] > 0)
      ]

Age Hospital Number  rectal temperature abdominal distension  \
0  Adult horse          534817                39.2               slight   
0  Adult horse          528548                38.1                 none   
0  Adult horse          534917                38.2                 none   
0  Adult horse          530526                38.3             moderate   
0  Adult horse          528653                37.8               slight   
0  Adult horse          530242                37.2               slight   
0  Adult horse          533847                37.8                 none   
0  Adult horse          528996                38.6               slight   
0  Adult horse          529642                37.2               slight   
0  Adult horse          529461                40.3             moderate   
0  Adult horse          529667                39.0               slight   
0  Adult horse          534784                37.9                 none   
0  Adult horse          529424                37.8                 none   
0  Adult horse          527518                36.4               slight   
0  Adult horse          534756                37.3               slight   
0  Adult horse          529703                39.0               slight   
0  Adult horse          529461                40.3             moderate   
0  Adult horse          530002                39.6             moderate   
0  Adult horse          527563                37.8                 none   
0  Adult horse          533696                38.7               severe   
0  Adult horse          534556                38.3                 none   
0  Adult horse          530360                38.2             moderate   
0  Adult horse          534572                38.0                 none   
0  Adult horse          529424                37.8                 none   
0  Adult horse          529399                39.3               slight   
0  Adult horse          528996                38.6               slight   
0  Adult horse          527365                38.4             moderate   
0  Adult horse         5299603                38.3                 none   
0  Adult horse          529628                38.5                 none   
0  Adult horse          528641                38.5             moderate   
0  Adult horse          534478                37.5               slight   
0  Adult horse          530401                37.9               slight   
0  Adult horse          529893                38.0               severe   
0  Adult horse          528183                37.6               slight   
0  Adult horse          528178                37.7               slight   
0  Adult horse          528931                38.4                 none   
0  Adult horse          535196                37.3             moderate   
0  Adult horse          530526                38.3             moderate   
0  Adult horse          534719                37.5               slight   
0  Adult horse          528743                37.1               slight   
0  Adult horse          530478                38.3             moderate   
0  Adult horse          534626                37.7                 none   
0  Adult horse          527702                37.2             moderate   
0  Adult horse          530612                36.5             moderate   

  rectal examination                            abdomen  \
0             absent                              other   
0          increased          distended large intestine   
0             absent          distended small intestine   
0          decreased          distended large intestine   
0             normal  firm feces in the large intestine   
0          decreased  firm feces in the large intestine   
0             normal                             normal   
0             normal  firm feces in the large intestine   
0          increased                             normal   
0             normal          distended large